File: ocaml/stats.ml
(* stats.ml *)

(* Calculate mean *)
let mean lst =
  let sum = List.fold_left (+) 0 lst in
  float_of_int sum /. float_of_int (List.length lst)

(* Calculate median *)
let median lst =
  let sorted = List.sort compare lst in
  let n = List.length sorted in
  if n mod 2 = 0 then
    let a = List.nth sorted (n/2 - 1) in
    let b = List.nth sorted (n/2) in
    (float_of_int (a + b)) /. 2.0
  else
    float_of_int (List.nth sorted (n/2))

(* Calculate mode *)
let mode lst =
  let freq_tbl =
    List.fold_left (fun acc x ->
      let count = try List.assoc x acc with Not_found -> 0 in
      (x, count + 1) :: List.remove_assoc x acc
    ) [] lst
  in
  let max_count =
    List.fold_left (fun m (_, c) -> if c > m then c else m) 0 freq_tbl
  in
  freq_tbl
  |> List.filter (fun (_, c) -> c = max_count)
  |> List.map fst, max_count

let () =
  let args = Array.to_list Sys.argv |> List.tl in
  let nums = List.map int_of_string args in
  if nums = [] then (
    Printf.printf "Usage: ocaml stats.ml <n1> <n2> ...\n";
    exit 1
  );
  let m = mean nums in
  let med = median nums in
  let modes, freq = mode nums in
  Printf.printf "Mean: %.2f\n" m;
  Printf.printf "Median: %.2f\n" med;
  Printf.printf "Mode (%d occurrences): %s\n"
    freq
    (String.concat ", " (List.map string_of_int modes))

File: c/stats.c
#include <stdio.h>
#include <stdlib.h>

int compare_ints(const void *a, const void *b) {
    int x = *(const int*)a;
    int y = *(const int*)b;
    return (x > y) - (x < y);
}

double mean(int arr[], int n) {
    long sum = 0;
    for (int i = 0; i < n; i++) sum += arr[i];
    return (double)sum / n;
}

double median(int arr[], int n) {
    qsort(arr, n, sizeof(int), compare_ints);
    if (n % 2 == 0)
        return (arr[n/2 - 1] + arr[n/2]) / 2.0;
    else
        return arr[n/2];
}

void mode(int arr[], int n, int **modes, int *mode_count, int *freq) {
    qsort(arr, n, sizeof(int), compare_ints);
    int max_count = 1, curr_count = 1;
    *mode_count = 0;
    *modes = malloc(n * sizeof(int));

    for (int i = 1; i < n; i++) {
        if (arr[i] == arr[i-1]) {
            curr_count++;
        } else {
            if (curr_count > max_count) {
                max_count = curr_count;
                *mode_count = 0;
                (*modes)[(*mode_count)++] = arr[i-1];
            } else if (curr_count == max_count) {
                (*modes)[(*mode_count)++] = arr[i-1];
            }
            curr_count = 1;
        }
    }
    // last run
    if (curr_count > max_count) {
        max_count = curr_count;
        *mode_count = 0;
        (*modes)[(*mode_count)++] = arr[n-1];
    } else if (curr_count == max_count) {
        (*modes)[(*mode_count)++] = arr[n-1];
    }
    *freq = max_count;
}

int main() {
    int n;
    printf("Enter number of elements: ");
    if (scanf("%d", &n) != 1 || n <= 0) return 1;

    int *arr = malloc(n * sizeof(int));
    printf("Enter %d integers:\n", n);
    for (int i = 0; i < n; i++) scanf("%d", &arr[i]);

    double m = mean(arr, n);
    double med = median(arr, n);

    int *modes;
    int mode_count, freq;
    mode(arr, n, &modes, &mode_count, &freq);

    printf("Mean: %.2f\n", m);
    printf("Median: %.2f\n", med);
    printf("Mode (%d occurrences): ", freq);
    for (int i = 0; i < mode_count; i++) {
        printf("%d ", modes[i]);
    }
    printf("\n");

    free(arr);
    free(modes);
    return 0;
}
